federation/build.sh

```diff
@@ -0,0 +1,211 @@
+readonly FEDERATION_OUTPUT_ROOT="${LOCAL_OUTPUT_ROOT}/federation"
+
+readonly KUBE_ANYWHERE_FEDERATION_IMAGE="gcr.io/madhusudancs-containers/kubernetes-anywhere-federation"
+readonly KUBE_ANYWHERE_FEDERATION_VERSION="v0.9.0"
+readonly KUBE_ANYWHERE_FEDERATION_CHARTS_IMAGE="gcr.io/madhusudancs-containers/federation-charts"
+readonly KUBE_ANYWHERE_FEDERATION_CHARTS_VERSION="v0.9.0"
+
+readonly GOOGLE_APPLICATION_CREDENTIALS="${GOOGLE_APPLICATION_CREDENTIALS:-${HOME}/.config/gcloud/application_default_credentials.json}"
+readonly KUBE_CONFIG_DIR="${KUBE_CONFIG_DIR:-${HOME}/.kube}"
+readonly KUBE_CONFIG="${KUBE_CONFIG:-${HOME}/.kube/config}"
+
+detect-project
+readonly KUBE_PROJECT="${KUBE_PROJECT:-${PROJECT:-}}"
+
+readonly KUBE_REGISTRY="${KUBE_REGISTRY:-gcr.io/${KUBE_PROJECT}}"
+# In dev environments this value must be recomputed after build. See
+# the build() function. Not making it readonly
+KUBE_VERSION="${KUBE_VERSION:-}"
+
+
+function cleanup {
+	rm -rf "${TMP_DIR}"
+	cd "${CUR_ROOT}"
+}
+trap cleanup EXIT
+
+function dirty_sha() {
+	local -r index="${KUBE_ROOT}/.git/index"
+	local -r objects_dir="${KUBE_ROOT}/.git/objects"
+
+	local -r tmp_dir="${TMP_DIR}/.git"
+	local -r tmp_index="${tmp_dir}/index"
+	local -r tmp_objects_dir="${tmp_dir}/objects"
+
+	mkdir -p "${tmp_objects_dir}"
+	cp "${index}" "${tmp_index}"
+
+	local -r files=$(git ls-files -m -o -d --exclude-standard)
+	GIT_INDEX_FILE="${tmp_index}" git add ${files}
+	GIT_ALTERNATE_OBJECT_DIRECTORIES="${objects_dir}" GIT_OBJECT_DIRECTORY="${tmp_objects_dir}" GIT_INDEX_FILE="${tmp_index}" git write-tree
+}
+
+function update_config() {
+	local -r q="${1:-}"
+	local -r cfile="${2:-}"
+	local -r bname="$(basename ${cfile})"
+
+	jq "${q}" "${cfile}" > "${TMP_DIR}/${bname}"
+	mv "${TMP_DIR}/${bname}" "${cfile}"
+}
+
+function build() {
+	kube::build::verify_prereqs
+	kube::build::build_image
+	kube::build::run_build_command make WHAT="cmd/kubectl cmd/hyperkube"
+
+	# Recompute KUBE_VERSION because it might have changed after rebuild.
+	KUBE_VERSION="${KUBE_VERSION:-$(kube::release::semantic_image_tag_version)}"
+
+	# Also append the dirty tree SHA to keep the versions unique across
+	# builds.
+	if [[ "${KUBE_VERSION}" == *-dirty ]]; then
+		KUBE_VERSION+=".$(dirty_sha)"
+	fi
+
+	BASEIMAGE="ubuntu:16.04" \
+		REGISTRY="${KUBE_REGISTRY}" \
+		VERSION="${KUBE_VERSION}" \
+		make -C "${KUBE_ROOT}/cluster/images/hyperkube" build
+}
+
+function push() {
+	kube::log::status "Pushing hyperkube image to the registry"
+	gcloud docker push "${KUBE_REGISTRY}/hyperkube-amd64:${KUBE_VERSION}"
+}
+
+function pull_installer() {
+	kube::log::status "Pulling installer images"
+	docker pull "${KUBE_ANYWHERE_FEDERATION_IMAGE}:${KUBE_ANYWHERE_FEDERATION_VERSION}"
+	docker pull "${KUBE_ANYWHERE_FEDERATION_CHARTS_IMAGE}:${KUBE_ANYWHERE_FEDERATION_CHARTS_VERSION}"
+}
+
+function ensure_files() {
+	kube::log::status "Ensure credential files exist..."
+	if [[ ! -f "${GOOGLE_APPLICATION_CREDENTIALS}" ]]; then
```

Comment:
Done. 

Also, a side note. `gce` is the only supported provider in this very first version. I want to get the first version in before I can improve it. I don't want to be continuously perfecting this in a vacuum. Improvements are going to land in follow up PRs.